File: java/spring-boot-mongodb/src/test/java/com/meridal/examples/springbootmongo/test/TestFramework.java
package com.meridal.examples.springbootmongo.test;

import com.meridal.examples.springbootmongo.domain.Recording;
import com.meridal.examples.springbootmongo.domain.Song;
import org.apache.commons.lang3.RandomStringUtils;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public abstract class TestFramework {

	private final Random random = new Random();

	protected Recording randomRecording() {
		return this.randomRecording(null);
	}

	protected Recording randomRecording(String id) {
		final Recording recording = new Recording();
		recording.setId(id);
		recording.setArtist(this.randomStringWithMaxLength(32));
		recording.setTitle(this.randomStringWithMaxLength(64));
		recording.setCatalogueNumber(this.randomStringWithMaxLength(16));
		recording.setYear(1965 + this.randomPositiveInteger(50));
		int tracks = this.randomPositiveInteger(20);
		final List<Song> songs = new ArrayList<>();
		for (int i = 0; i < tracks; i++) {
			songs.add(this.randomSong());
		}
		recording.setSongs(songs);
		return recording;
	}

	protected Recording cloneRecording(final String id, final Recording recording) {
		final Recording clone = new Recording();
		clone.setId(id);
		clone.setArtist(recording.getArtist());
		clone.setTitle(recording.getTitle());
		clone.setYear(recording.getYear());
		clone.setCatalogueNumber(recording.getCatalogueNumber());
		final List<Song> songs = new ArrayList<>(recording.getSongs());
		clone.setSongs(songs);
		return clone;
	}

	protected Song randomSong() {
		final Song song = new Song();
		song.setTitle(this.randomStringWithMaxLength(64));
		String mins = Integer.toString(this.randomPositiveInteger(20));
		String secs = Integer.toString(this.randomPositiveInteger(59));
		song.setDuration(mins + ":" + secs);
		return song;
	}

	protected Song cloneSong(final Song song) {
		final Song clone = new Song();
		clone.setDuration(song.getDuration());
		clone.setTitle(song.getTitle());
		return clone;
	}

	private int randomPositiveInteger(int max) {
		return Math.floorMod(random.nextInt(), max - 1) + 1;
	}

	private String randomStringWithMaxLength(int max) {
		return RandomStringUtils.randomAlphanumeric(1, max);
	}
}

File: kotlin/kotlin-spring-boot-dynamodb/src/test/kotlin/com/meridal/examples/kotlinspringboot/test/TestFramework.kt
package com.meridal.examples.kotlinspringboot.test

import com.meridal.examples.kotlinspringboot.model.Recording
import com.meridal.examples.kotlinspringboot.model.Song
import org.apache.commons.lang3.RandomStringUtils
import java.util.*

abstract class TestFramework {

    private val random = Random()

    protected fun randomRecording() = randomRecording(null)

    protected fun randomRecording(id: String?): Recording {
        val recording = Recording()
        recording.id = id
        recording.artist = randomStringWithMaxLength(32)
        recording.title = randomStringWithMaxLength(64)
        recording.catalogueNumber = randomStringWithMaxLength(16)
        recording.year = 1965 + randomPositiveInteger(50)
        val songs: MutableList<Song> = mutableListOf()
        for (i in 1..randomPositiveInteger(20)) songs += randomSong()
        recording.songs = songs
        return recording
    }

    protected fun cloneRecording(id: String?, recording: Recording): Recording {
        val clone = Recording()
        clone.id = id
        clone.artist = recording.artist
        clone.title = recording.title
        clone.catalogueNumber = recording.catalogueNumber
        clone.year = recording.year
        clone.songs = recording.songs
        return clone
    }

    protected fun randomSong(): Song {
        val song = Song()
        song.title = randomStringWithMaxLength(64)
        val minutes = randomPositiveInteger(22)
        val seconds = randomPositiveInteger(60) - 1
        song.duration = "$minutes:$seconds"
        return song
    }

    protected fun cloneSong(id: Long, song: Song): Song {
        val clone = Song()
        clone.title = song.title
        clone.duration = song.duration
        return clone
    }

    private fun randomPositiveInteger(max: Int) = Math.floorMod(random.nextInt(), max - 1)

    private fun randomStringWithMaxLength(max: Int) = RandomStringUtils.randomAlphanumeric(1, max)
}